src/mlpack/methods/ann/activation_functions/isru_function.hpp

```diff
@@ -0,0 +1,137 @@
+ 
+ #include <mlpack/prereqs.hpp>
+ 
+namespace mlpack{
+namespace ann /** Artifical Neural Network. */{
+  
+/**
+ * The ISRU function, defined by
+ * 
+ * @f{eqnarray*}{
+ * f(x) &=& x/sqrt(1 + alpha*x*x)
+ * f'(x) &=& cube(1/sqrt(1 + alpha*x*x))
+ * @f
+ */
+
+class ISRUFunction
+{
+  
+ public:
+  /**
+   * Computes the ISRU function
+   * 
+   * @param x Input data.
+   * @param alpha paramter, default value = 0.1
+   * @return f(x)
+   */
+  static double Fn(const double x, const double alpha = 0.1)
+  {
+    return x / (std::sqrt(1 + alpha*x*x));
+  }
+  
+  /**
+   * Computes the ISRU function using a matrix as a input
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename eT>
+  static void Fn(const arma::Mat<eT>& x,
+                 arma::Mat<eT>& y, 
+                 const double alpha = 0.1)
+  {
+    y = x / (arma::sqrt(1 + alpha * arma::pow(x, 2)));
+  }
+
+  /**
+   * Computes the ISRU function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, 
+                 OutputVecType& y, 
+                 const double alpha = 0.1)
+  {
+    y.set_size(arma::size(x));
+
+    for (size_t i = 0; i < x.n_elem; i++)
+      y(i) = Fn(x(i));
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y
+   * @param alpha
+   * @return f^{-1}(y)
+   */
+  static double Inv(const double y, const double alpha = 0.1)
+  {
+    return y / std::sqrt(1 - alpha * y * y);
+  }
+  
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y Input data.
+   * @param x The resulting inverse of the input data
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y,
+                  OutputVecType& x, 
+                  const double alpha = 0.1)
+  {
+    x = y / arma::sqrt(1 - alpha * arma::pow(y, 2 ));
+  }
+  
+  /**
+   * Computes the first derivate of the ISRU function
+   * 
+   * @param y Input activation
+   * @param alpha parameter, default value = 0.1
+   * @return f'(x) where f(x) = y
+   */
+  static double Deriv(const double y, const double alpha = 0.1)
+  {
+    double x = Inv(y, alpha);
+    return std::pow(y / x, 3); 
+  }
+
+  /**
+   * Computes the first derivative of the ISRU function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y,
+                    OutputVecType& x, 
+                    const double alpha = 0.1)
+  {
+    x = arma::pow(y / x, 3);
```

Comment:
you need to calculate x = Inv(y, alpha) here. Personally I would prefer it if you reduced the expression so that you directly assign value rather than calculating inverse.

Reply:
I believe this is causing Ctest failures.

Reply:
I think it would be better to add a fatal error in case of division by zero.

Reply:
> I think it would be better to add a fatal error in case of division by zero.

While calculating the inverse of this function, the value gets undefined(in case of division by zero) only when y = 1/alpha^{1/2}, while the range of y = f(x) is (-1/alpha^{1/2} , 1/alpha^{1/2}).
So theoretically that case will never be called however maybe due to precision it can occur. Then would it better to just set y  = 1/alpha^{1/2} - epsilon (say 1e-6) if y >= 1/alpha^{1/2} and set y  = -1/alpha^{1/2} + epsilon if                       y <= -1/alpha^{1/2}. Or should a fatal error be thrown(like you suggested)?

Reply:
I think rather than changing definition to subtract 1e-6, lets throw an error if needed.However I think a member would respond better. I wanted to do so because one might only call derivatives of y for any input. I will take a closer look at the math, if that case is unlikely we shall leave it as it is.

Reply:
Ok. Thanks for your input. If you feel like it is necessary to detect those edge cases I'll add check to it. After you review it tomorrow please tell me if anything else should be done.